vm/compiler/src/coinbase_puzzle/mod.rs

```diff
@@ -0,0 +1,363 @@
+            "Invalid number of challenge points"
+        );
+
+        // Pop the last challenge point as the accumulator challenge point.
+        let accumulator_point = match challenge_points.pop() {
+            Some(point) => point,
+            None => bail!("Missing the accumulator challenge point"),
+        };
+
+        // Compute the accumulator evaluation.
+        let mut accumulator_evaluation = cfg_iter!(prover_polynomials)
+            .zip_eq(&challenge_points)
+            .fold(<N::PairingCurve as PairingEngine>::Fr::zero, |accumulator, (prover_polynomial, challenge_point)| {
+                accumulator + (prover_polynomial.evaluate(accumulator_point) * challenge_point)
+            })
+            .sum();
+        accumulator_evaluation *= &epoch_challenge.epoch_polynomial().evaluate(accumulator_point);
+
+        // Compute the accumulator commitment.
+        let commitments: Vec<_> =
+            cfg_iter!(coinbase_solution.partial_solutions()).map(|solution| solution.commitment().0).collect();
+        let fs_challenges = challenge_points.into_iter().map(|f| f.to_repr()).collect::<Vec<_>>();
+        let accumulator_commitment =
+            KZGCommitment::<N::PairingCurve>(VariableBase::msm(&commitments, &fs_challenges).into());
+
+        // Retrieve the coinbase verifying key.
+        let coinbase_verifying_key = match self {
+            Self::Prover(coinbase_proving_key) => &coinbase_proving_key.verifying_key,
+            Self::Verifier(coinbase_verifying_key) => coinbase_verifying_key,
+        };
+
+        // Return the verification result.
+        Ok(KZG10::check(
+            coinbase_verifying_key,
+            &accumulator_commitment,
+            accumulator_point,
+            accumulator_evaluation,
+            coinbase_solution.proof(),
+        )?)
+    }
+
+    /// Returns the coinbase proving key.
+    pub fn coinbase_proving_key(&self) -> Result<&CoinbaseProvingKey<N>> {
+        match self {
+            Self::Prover(coinbase_proving_key) => Ok(coinbase_proving_key),
+            Self::Verifier(_) => bail!("Cannot fetch the coinbase proving key with a verifier"),
+        }
+    }
+
+    /// Returns the coinbase verifying key.
+    pub fn coinbase_verifying_key(&self) -> Result<&CoinbaseVerifyingKey<N>> {
+        match self {
+            Self::Prover(coinbase_proving_key) => Ok(&coinbase_proving_key.verifying_key),
+            Self::Verifier(coinbase_verifying_key) => Ok(coinbase_verifying_key),
+        }
+    }
+}
+
+impl<N: Network> CoinbasePuzzle<N> {
+    /// Returns the prover polynomial for the coinbase puzzle.
+    fn prover_polynomial(
+        epoch_challenge: &EpochChallenge<N>,
+        address: &Address<N>,
+        nonce: u64,
+    ) -> Result<DensePolynomial<<N::PairingCurve as PairingEngine>::Fr>> {
+        let input = {
+            let mut bytes = [0u8; 80];
+            bytes[..40].copy_from_slice(&epoch_challenge.to_bytes_le()?);
```

Comment:
in the absence of a related `const` it might make sense to at least preface this operation with a `debug_assert` that the length of the serialized `epoch_challenge` is `40`